### ISSUE TITLE
Tracking Issue for merging to upstream

### ISSUE THREAD

**Issue by polarathene** [espressif/llvm-project#4]
I understand it will take quite some time for this backend to be merged upstream into LLVM. 

This issue is for notifying anyone that subscribes to it when the merge to upstream has completed closing this issue.

**Comment by igrr** [comment]
We are moving this and other issues to https://github.com/espressif/llvm-project, which will host the Xtensa patches for LLVM based on the new monorepo layout. This new repository will be updated with Xtensa patches based on LLVM 9 soon.

**Comment by andreisfr** [comment]
@polarathene, @DarkWiiPlayer , @breckwagner,  we updated current patches (1-10), which are in LLVM review, according to changes in upstream version and reviewers requests for changes.

**Comment by lexxvir** [comment]
Hi @andreisfr,

Is there any plan/way to merge this fork into the upstream?

I've skimmed some pull requests on the reviews.llvm.org, it seems the process is stalled.

**Comment by andreisfr** [comment]
Hi @lexxvir,

Yes, we plan to continue integration process of the Xtensa backend to the upstream.

**Comment by windoze** [comment]
Is there any progress for this? 
I'm dying to see it merging into upstream, it's really a PITA to build the whole LLVM/Clang project from the scratch in a CI pipeline.

**Comment by shymega** [comment]
Hey, hope you don't mind me chasing this up :smile: I've been getting ready to start work on Rust code for an ESP32, but without official LLVM and Rust support, I fear I'll have to use C. It does take ages to compile as well. I might be able to influence a different chip, but I think the ESP32 is a solid choice.

Thanks.

**Comment by torpak** [comment]
It seems nobody at LLVM is even answering since may. Does anyone know a reason for this?

**Comment by clouds56** [comment]
@torpak according to this https://reviews.llvm.org/D64830#2615149 comment, I think they are waiting the author to rewrite a RFC since things changed a lot.
Correct me if wrong.

**Comment by Hao123br** [comment]
@clouds56 the author already sent another RFC. https://lists.llvm.org/pipermail/llvm-dev/2021-March/149090.html

**Comment by clouds56** [comment]
@Hao123br thanks for the link, shall we also provide the link in the thread and ping aykevl & tstellar, maybe they are too busy and missing the RFC?

**Comment by saeidakbari** [comment]
Is any new news on reviewing process?

**Comment by flaminisx** [comment]
just noticed that patches 4, 5, 6, 8 and 10 are also accepted

**Comment by igrr** [comment]
For future reference, the new RFC is at https://discourse.llvm.org/t/rfc-request-for-upstream-tensilica-xtensa-esp32-backend/65355.

**Comment by SergioGasquez** [comment]
> > The patches are in review (see links in https://esp32.com/viewtopic.php?t=11412#p46446) and further timeline will depend on the review process. We will keep updating the patch set to address the feedback!
> 
> From Espressif Forum post [Re: Status of LLVM support?](https://esp32.com/viewtopic.php?t=11412#p46446)
> 
> > * [x]  [D64826: [Xtensa 1/10] Recognize Xtensa in triple parsing code.](https://reviews.llvm.org/D64826)
> > * [x]  [D64827: [Xtensa 2/10] Add Xtensa ELF definitions.](https://reviews.llvm.org/D64827)
> > * [x]  [D64829: [Xtensa 3/10] Add initial version of the Xtensa backend.](https://reviews.llvm.org/D64829)
> > * [x]  [D64830: [Xtensa 4/10] Add basic *td files with Xtensa architecture description.](https://reviews.llvm.org/D64830)
> > * [x]  [D64831: [Xtensa 5/10] Add Xtensa MCTargetDescr initial functionality.](https://reviews.llvm.org/D64831)
> > * [x]  [D64832: [Xtensa 6/10] Add Xtensa basic assembler parser.](https://reviews.llvm.org/D64832)
> > * [ ]  [D64833: [Xtensa 7/10] Add Xtensa instruction printer.](https://reviews.llvm.org/D64833)
> > * [x]  [D64834: [Xtensa 8/10] Add support of the Xtensa shift/load/store/move and processor control instructions.](https://reviews.llvm.org/D64834)
> > * [ ]  [D64835: [Xtensa 9/10] Add basic support of Xtensa disassembler.](https://reviews.llvm.org/D64835)
> > * [x]  [D64836: [Xtensa 10/10] Add relaxations and fixups. Add rest part of Xtensa Core Instructions.](https://reviews.llvm.org/D64836)
> 
> Status as of 2019-12-24 [x] = Accepted [ ] = Needs Review

Patches 7 and 9 have now been accepted!

**Comment by shish** [comment]
Looks like these are all merged into the main branch now, if I'm reading it right? Merry Christmas all :D

**Comment by aq1018** [comment]
Should this be closed? LLVM has has accepted all patches according to this article - https://www.phoronix.com/news/LLVM-Xtensa-Backend

**Comment by Skgland** [comment]
Unless https://github.com/esp-rs/rust/issues/89#issuecomment-1426982176 is outdated, there are still some parts missing.

**Comment by igrr** [comment]
The next set of patches has been uploaded to LLVM Phabricator. I am adding the list of patches in the same format as before. We'll be updating this comment as patches are accepted.

* [ ] [D145658: [Xtensa] Initial support of the ALU operations.](https://reviews.llvm.org/D145658)
* [ ] [D145660: [Xtensa] Codegen support for memory operations](https://reviews.llvm.org/D145660)
* [ ] [D145661: [Xtensa] Add Constant Pool](https://reviews.llvm.org/D145661)
* [ ] [D145662: [Xtensa] Implement asm representation of the Constant Pool.](https://reviews.llvm.org/D145662)
* [ ] [D145663: [Xtensa] Implement lowering constants.](https://reviews.llvm.org/D145663)
* [ ] [D145664: [Xtensa] Add support of the Xtensa function calls](https://reviews.llvm.org/D145664)
* [ ] [D145665: [Xtensa] Implement lowering ConstantPool and address operations.](https://reviews.llvm.org/D145665)
* [ ] [D145666: [Xtensa] Implement emitPrologue/emitEpilogue](https://reviews.llvm.org/D145666)
* [ ] [D145667: [Xtensa] Lower stack operations.](https://reviews.llvm.org/D145667)
* [ ] [D145668: [Xtensa] Implement lowering SELECT\_CC, SETCC](https://reviews.llvm.org/D145668)
* [ ] [D146948: [Xtensa] Support for a variety of additional LLVM IR constructs.](https://reviews.llvm.org/D146948)
* [ ] [D146949: [Xtensa] Lower SHIFT PARTS and shift operations.](https://reviews.llvm.org/D146949)
* [ ] [D146950: [Xtensa] Implement load pseudo operations and patterns.](https://reviews.llvm.org/D146950)
* [ ] [D146951: [Xtensa] Support for variable arguments](https://reviews.llvm.org/D146951)
* [ ] [D146952: [Xtensa] Implement lowering BR\_JT operation](https://reviews.llvm.org/D146952)
* [ ] [D146953: [Xtensa] Support for address intrinsics.](https://reviews.llvm.org/D146953)
* [ ] [D146954: [Xtensa] Add basic support for inline asm constraints](https://reviews.llvm.org/D146954)
* [ ] [D146955: [Xtensa] Implement volatile load/store.](https://reviews.llvm.org/D146955)
* [ ] [D146956: [Xtensa] Implement branch analysis](https://reviews.llvm.org/D146956)
* [ ] [D146957: [Xtensa] Implement support for the BranchRelaxation](https://reviews.llvm.org/D146957)

**Comment by orowith2os** [comment]
@igrr @andreisfr reading some of the reviews, a common occurrence is the lack of tests. How's that going? It's been a few months, so figured I'd ask.

**Comment by andreisfr** [comment]
@orowith2os , we preparing update for patches, including tests, soon it will be uploaded to Phabricator.

**Comment by Skgland** [comment]
It appears that the llvm-project is movin/has moved from Phabricator to GitHub (<https://github.com/llvm/llvm-project/pull/65162>) and Phabricator is suppost to go read-only next month/in October. I guess the patches need to be resubmitted via GitHub at some point.

**Comment by teburd** [comment]
Very cool, definitely looking forward to seeing this make its way in

**Comment by alhirzel** [comment]
For those keeping track, it looks like [a recently-merged PR](https://github.com/llvm/llvm-project/pull/78548) covers the first five of the above-mentioned patchsets:

> * [x]  [D145658: [Xtensa] Initial support of the ALU operations.](https://reviews.llvm.org/D145658)
> * [x]  [D145660: [Xtensa] Codegen support for memory operations](https://reviews.llvm.org/D145660)
> * [x]  [D145661: [Xtensa] Add Constant Pool](https://reviews.llvm.org/D145661)
> * [x]  [D145662: [Xtensa] Implement asm representation of the Constant Pool.](https://reviews.llvm.org/D145662)
> * [x]  [D145663: [Xtensa] Implement lowering constants.](https://reviews.llvm.org/D145663)
>  * [ ]  [D145664: [Xtensa] Add support of the Xtensa function calls](https://reviews.llvm.org/D145664)
>  * [ ]  [D145665: [Xtensa] Implement lowering ConstantPool and address operations.](https://reviews.llvm.org/D145665)
>  * [ ]  [D145666: [Xtensa] Implement emitPrologue/emitEpilogue](https://reviews.llvm.org/D145666)
>  * [ ]  [D145667: [Xtensa] Lower stack operations.](https://reviews.llvm.org/D145667)
>  * [ ]  [D145668: [Xtensa] Implement lowering SELECT_CC, SETCC](https://reviews.llvm.org/D145668)
>  * [ ]  [D146948: [Xtensa] Support for a variety of additional LLVM IR constructs.](https://reviews.llvm.org/D146948)
>  * [ ]  [D146949: [Xtensa] Lower SHIFT PARTS and shift operations.](https://reviews.llvm.org/D146949)
>  * [ ]  [D146950: [Xtensa] Implement load pseudo operations and patterns.](https://reviews.llvm.org/D146950)
>  * [ ]  [D146951: [Xtensa] Support for variable arguments](https://reviews.llvm.org/D146951)
>  * [ ]  [D146952: [Xtensa] Implement lowering BR_JT operation](https://reviews.llvm.org/D146952)
>  * [ ]  [D146953: [Xtensa] Support for address intrinsics.](https://reviews.llvm.org/D146953)
>  * [ ]  [D146954: [Xtensa] Add basic support for inline asm constraints](https://reviews.llvm.org/D146954)
>  * [ ]  [D146955: [Xtensa] Implement volatile load/store.](https://reviews.llvm.org/D146955)
>  * [ ]  [D146956: [Xtensa] Implement branch analysis](https://reviews.llvm.org/D146956)
>  * [ ]  [D146957: [Xtensa] Implement support for the BranchRelaxation](https://reviews.llvm.org/D146957)

**Comment by andreisfr** [comment]
@alhirzel ,  recent  PR contains only 1st mentioned patch. We removed 4 patches from PR, according to reviewers comments. But we preparing next patch now)

**Comment by alhirzel** [comment]
> @alhirzel , recent PR contains only 1st mentioned patch. We removed 4 patches from PR, according to reviewers comments. But we preparing next patch now)

Gotcha - edited to correct this. Sorry!

**Comment by gerekon** [comment]
Hi guys!

LLVM upstreaming process has moved to GitHub and looks like original patches posted on Phabricator are going to be restructured. I suppose that in the most cases they will be split into pieces.

We decided to avoid posting a lot of patches at the same time in order to ease review and update process. So my suggestion is to update the list and include only patches which has respective PR on GitHub. So the actual list should look like below. And we will update it with new PRs.

- [x] [[Xtensa] Initial codegen support from IR](https://github.com/llvm/llvm-project/pull/78548). (Part of [D145658: [Xtensa] Initial support of the ALU operations.](https://reviews.llvm.org/D145658))

@andreisfr @sstefan1 WDYT?

**Comment by andreisfr** [comment]
@gerekon, I think yes, the new series of patches will be different from the previous one

**Comment by gerekon** [comment]
@andreisfr Could you mention this PR in every upstream PR? In order to make tracking easier

**Comment by andreisfr** [comment]
> @andreisfr Could you mention this PR in every upstream PR? In order to make tracking easier

yes

**Comment by gerekon** [comment]
Upstream patches list update
- [x] [[Xtensa] Initial codegen support from IR](https://github.com/llvm/llvm-project/pull/78548). (Split from [D145658: [Xtensa] Initial support of the ALU operations.](https://reviews.llvm.org/D145658))
- [ ] [[Xtensa] Implement base CallConvention](https://github.com/llvm/llvm-project/pull/83280).

**Comment by SergioGasquez** [comment]
https://github.com/llvm/llvm-project/pull/83280 is already merged! :tada:

**Comment by gerekon** [comment]
:smiley: 
Upstream patches list update
- [x] [[Xtensa] Initial codegen support from IR](https://github.com/llvm/llvm-project/pull/78548). (Split from [D145658: [Xtensa] Initial support of the ALU operations.](https://reviews.llvm.org/D145658))
- [x] [[Xtensa] Implement base CallConvention](https://github.com/llvm/llvm-project/pull/83280).

**Comment by gerekon** [comment]
Upstream patches list update. We pushed new patch to upstream
- [x] [[Xtensa] Initial codegen support from IR](https://github.com/llvm/llvm-project/pull/78548). (Split from [D145658: [Xtensa] Initial support of the ALU operations.](https://reviews.llvm.org/D145658))
- [x] [[Xtensa] Implement base CallConvention](https://github.com/llvm/llvm-project/pull/83280).
- [ ] [[Xtensa] Implement FrameLowering methods and stack operation lowering](https://github.com/llvm/llvm-project/pull/92960)

**Comment by andreisfr** [comment]
Upstream patches list update

- [ ] https://github.com/llvm/llvm-project/pull/78548. (Split from [D145658: [Xtensa] Initial support of the ALU operations.](https://reviews.llvm.org/D145658))
- [ ]  https://github.com/llvm/llvm-project/pull/83280.
- [ ]  https://github.com/llvm/llvm-project/pull/92960
- [ ] https://github.com/llvm/llvm-project/pull/95256

**Comment by tom-borcin** [comment]
@SergioGasquez There are 13 patches left:

1. Implement lowering SELECT_CC, SETCC  - In Progress
2. Lower stack operations.
3. Support for a variety of additional LLVM IR constructs.
4. Lower SHIFT PARTS and shift operations.
5. Implement load pseudo operations and patterns.
6. Support for variable arguments
7. Implement lowering BR_JT operation
8. Support for address intrinsics.
9. Add basic support for inline asm constraints
10. Implement volatile load/store.
11. Implement branch analysis
12. Implement support for the BranchRelaxation

**Comment by gerekon** [comment]
@tom-borcin These are the patches initially posted to Phabricator. I think that was just some part of our patches.  After LLVM development moved to GH the structure of patches has changed so the list above is not correct.
I think it is better to ask @andreisfr about the list of patches.

**Comment by andreisfr** [comment]
@SergioGasquez , @tom-borcin , 
Probably we need to merge about 30-40 patches to implement most of functionality to be able generate Rust code and compile ESP-IDF examples:
1) Currently we have 10 prepared patches with base functionality not upstreamed (https://github.com/espressif/llvm-project/issues/4#issuecomment-2189374398). This patches are not very large, so I think optimistic estimation is to spend 1-1,5 week on each patch to upstream. 
So process in optimal way will be finished in September. 
2) Implement other Xtensa features like Window, Boolean, Float, etc. It is about 8-12 patches. Also in optimistic case will take another 3 month and wiil be finished in November. 
3) Implement Xtensa targets in Clang esp32,esp8266, etc" and add support for multilib and linker. It is about 8-12 patches. Some of these patches will be complex, so in optimistic case will take another 2-3 month and will be finished in January/Febraury. 
In a pessimistic case, this process may take another 2-4 months.

**Comment by JetForMe** [comment]
Pity it can't be sooner! What's the main obstacle? Getting upstream to accept the patches?

**Comment by nlgranger** [comment]
I would hazard a guess that it's about upstream finding time to review the patches.

**Comment by andreisfr** [comment]
- [ ] 1. https://github.com/llvm/llvm-project/pull/78548. (Split from [D145658: [Xtensa] Initial support of the ALU operations.]
- [ ] 2. https://reviews.llvm.org/D145658))
- [ ] 3.  https://github.com/llvm/llvm-project/pull/83280.
- [ ] 4.  https://github.com/llvm/llvm-project/pull/92960
- [ ] 5.  https://github.com/llvm/llvm-project/pull/95256
- [ ] 6. https://github.com/llvm/llvm-project/pull/97017

**Comment by andreisfr** [comment]
@JetForMe , @nlgranger delay in pull requests merging is that we must to update commits according to reviewers comments,  this is iterative process and also currently only a few reviewers commenting our PRs. Maybe, If you have the opportunity to involve more code reviewers, that would be great.

**Comment by andreisfr** [comment]
@bugadani , we have about 10-15 patches, which depends on each other. I think,  after merging these patches we will be able to create several PR at the same time.

**Comment by aykevl** [comment]
@andreisfr one way to interest more people in reviewing Xtensa patches might be to review patches from other people. Not sure if you are doing this already, but it is usually very much appreciated. When they see you reviewing their code, they might be more likely to review Xtensa patches.

**Comment by andreisfr** [comment]
@aykevl , thank you for advice!

**Comment by JetForMe** [comment]
I will try to bring this to the attention of the Embedded Swift folks at Apple. They currently support the ESP32-C6, and I think very much want to encourage Matter adoption, so convincing them that Xtensa support is important might help speed things up.

I suspect Espressif has their own connections to Apple, at the level of the Matter standards development. It would be good to shake that tree, too.

**Comment by mortenbekditlevsen** [comment]
@JetForMe Will you do this through the Swift Evolution Forum@

@andreisfr In case you're wondering about the spike in interest for these patches, the Swift programming language quite recently added an 'embedded' subset of the language. 
This is why myself and many others are now following this work with much interest. 😊

**Comment by MihaelIsaev** [comment]
I'm subscribed to all the activity in this repository because waiting for Xtensa support to start using ESP32 with Embedded Swift. Would love to support a thread about that on Swift Forums 😊 there are [already at least one thread](https://forums.swift.org/t/is-there-support-for-xtensa-lx7-cores/72458) and [on reddit](https://www.reddit.com/r/esp32/comments/1ddjbbk/embedded_swift_on_esp32/) as well

**Comment by teburd** [comment]
Are the hifi patches also upstreamed?

**Comment by gerekon** [comment]
@teburd 
> Are the hifi patches also upstreamed?

Do you mean ones from [this PR](https://github.com/espressif/llvm-project/pull/85)?
Not yet I think. They need Boolean and Floating point features as far as I see.

**Comment by bugadani** [comment]
Considering how long it takes to upstream few-hundered-LOC PRs, I am somewhat worried about throwing 80k+ LOC patches in 😅

**Comment by gerekon** [comment]
> Considering how long it takes to upstream few-hundered-LOC PRs, I am somewhat worried about throwing 80k+ LOC patches in 😅

I expect that after the next 10-15 patches we will start doing this in parallel. Maybe we can try to keep 2-3 active patches in upstream review and see if it will speed up the process.

@andreisfr what do you think?

**Comment by andreisfr** [comment]
@gerekon , yes, I think we need to upstream about 10-15 patches and probably we will be able to create parallel PRs

**Comment by ProfFan** [comment]
Is there an update or target timeline?

**Comment by johnnyasantoss** [comment]
Apparently https://github.com/llvm/llvm-project/pull/99981 just got merged

**Comment by JetForMe** [comment]
IIUC everything is merged? I wonder how long before it makes it into a Swift toolchain.

**Comment by Skgland** [comment]
> IIUC everything is merged?

No, https://github.com/espressif/llvm-project/issues/4#issuecomment-2260035936 only includes the seven patches already submitted to upstream. Based on https://github.com/espressif/llvm-project/issues/4#issuecomment-2201145347 that makes it 7 out of 10-15 patches that need to be submitted in series after which more patches will follow in parallel.

**Comment by JetForMe** [comment]
[This comment](https://github.com/espressif/llvm-project/issues/4#issuecomment-2189374398) listed the following items remain. I was unable to easily correlate them to patches in later comments.

> @SergioGasquez There are 13 patches left:
> 
> 1. Implement lowering SELECT_CC, SETCC  - In Progress
> 2. Add support of the Xtensa function calls
> 3. Lower stack operations.
> 4. Support for a variety of additional LLVM IR constructs.
> 5. Lower SHIFT PARTS and shift operations.
> 6. Implement load pseudo operations and patterns.
> 7. Support for variable arguments
> 8. Implement lowering BR_JT operation
> 9. Support for address intrinsics.
> 10. Add basic support for inline asm constraints
> 11. Implement volatile load/store.
> 12. Implement branch analysis
> 13. Implement support for the BranchRelaxation

**Comment by andreisfr** [comment]
Current status. 

- [ ]  1. https://github.com/llvm/llvm-project/pull/78548. (Split from [D145658: [Xtensa] Initial support of the ALU operations.]
- [ ]  2. https://reviews.llvm.org/D145658))
- [ ]  3. https://github.com/llvm/llvm-project/pull/83280.
- [ ]  4. https://github.com/llvm/llvm-project/pull/92960
- [ ]  5. https://github.com/llvm/llvm-project/pull/95256
- [ ]  6. https://github.com/llvm/llvm-project/pull/97017
- [ ]  7. https://github.com/llvm/llvm-project/pull/99981
- [ ]  8. https://github.com/llvm/llvm-project/pull/106053

**Comment by TurkeyMan** [comment]
🎉 

What happens next now that these all seem to be done?

**Comment by bugadani** [comment]
Now the next patch will be submitted, then the next, and the next... :)

**Comment by TurkeyMan** [comment]
Lol, but as someone following closely and with much excitement... it seems to be the end of the list of outstanding patches that I was aware of. Wasn't clear if there's another whole bunch, or if we're approaching branch synchronisation?

**Comment by bugadani** [comment]
Considering Andrej has estimated 10-15 patches two months ago to be submitted one by one, and since then 2 got merged, I wouldn't hold my breath. And as I understand even that isn't the end of this story.

**Comment by andreisfr** [comment]
The https://github.com/llvm/llvm-project/pull/106053 PR is merged.
We preparing next PR now.

**Comment by Skgland** [comment]
Maybe the `Current status` Comment could include an additional bullet point at the and for forther patches to avoid this confusion.

**Comment by andreisfr** [comment]
> Maybe the `Current status` Comment could include an additional bullet point at the and for further patches to avoid this confusion.

Ok, I will add this information in next update

**Comment by andreisfr** [comment]
Hi all!

I prepared a plan which describes nearest steps of upstreaming Xtensa backend. The goal is to implement most of functionality to be able generate Rust code and compile ESP-IDF examples.


1) We have 6 prepared patches with base functionality not upstreamed. This patches are not very large, so I think optimistic estimation is to finish this step till middle of October.

- [ ] https://github.com/llvm/llvm-project/pull/107363
- [ ] Add basic support for inline asm constraints
- [ ] Implement volatile load/store.
- [ ] Implement branch analysis
- [ ] Implement support for the BranchRelaxation
- [ ] Support for variable arguments

 
2) Implement other Xtensa features like Window, Boolean, Float, etc. At this step we didn't prepare exactly patch list, because it depends on previous step, it will be about 8-12 patches. Also in optimistic case will take another 2-3 month and probably be finished in December. At this stage we probably can make some parallel PRs.

Currently we have feature list to implement:
- Implement code density feature 
- Implement Windowed feature 
- Implement FP feature 
- Implement Boolean feature 
- Implement Mac16 feature and operations. 
- Implement Debug, DFPAccel, S32C1I, THREADPTR, Extended L32R, ATOMCTL, MEMCTL features.
- Implement Exception, HighPriInterrupts, Coprocessor, Interrupt, RelocatableVector, TimerInt, PRID, RegionProtection and 
- MiscSR  features. Implement instructions for Exception, Interrupt and RegionProtection features.

We also have a patches for this features, but number of patches and order maybe changed. I will prepare more precise patch set when first step will be finished.

3) Implement Xtensa targets in Clang esp32,esp8266, etc" and add support for multilib and linker. It is about 8-12 patches. Some of these patches will be complex, so in optimistic case will take another 2-3 month and will be finished in Febraury/March.  At this stage we probably can make some parallel PRs.
After that we can probably build esp-idf and Rust examples.

4) Implement minor features, like hwloop, etc.

**Comment by andreisfr** [comment]
- [ ] https://github.com/llvm/llvm-project/pull/107363
- [ ] https://github.com/llvm/llvm-project/pull/108986
- [ ]  Implement volatile load/store.
- [ ]  Implement branch analysis
- [ ]  Implement support for the BranchRelaxation
- [ ]  Support for variable arguments

**Comment by andreisfr** [comment]
The Xtensa backend upstreaming status update:

- [ ]  https://github.com/llvm/llvm-project/pull/107363
- [ ]  https://github.com/llvm/llvm-project/pull/108986
- [ ]  Implement volatile load/store.
- [ ]  [Implement branch analysis](https://github.com/llvm/llvm-project/pull/110959)
- [ ]  [Implement support for the BranchRelaxation](https://github.com/llvm/llvm-project/pull/110959)
- [ ]  Support for variable arguments

**Comment by andreisfr** [comment]
The Xtens backend upstreaming status update. Currently we preparing PR with the last patch from step 1 https://github.com/espressif/llvm-project/issues/4#issuecomment-2343204459 . In a few days I will an update with a list of the next patch set for upstream.

**Comment by gerekon** [comment]
Hi guys one more PR is pending which implements generic Xtensa target support in Clang. It is related to the next (2nd) set of patches mentioned [here](https://github.com/espressif/llvm-project/issues/4#issuecomment-2343204459).

### Backend base implementation
 - [ ] [Support for variable arguments](https://github.com/llvm/llvm-project/pull/117126)
 
### Xtensa features and toolchain implementation

LLVM:
- [ ] Implement Xtensa density option
- [ ] Implement Xtensa windowed option
- [ ] Implement Xtensa FP option and operations
- [ ] Implement Loop, SEXT and NSA features.
- [ ] Implement Mul32, Mul32High and Div32 features.
- [ ] Implement Mac16 feature and operations.
- [ ] Implement Debug, DFPAccel, S32C1I, THREADPTR, Extended L32R, ATOMCTL, MEMCTL features.
- [ ] Implement Exception, HighPriInterrupts, Coprocessor, Interrupt, RelocatableVector, TimerInt, PRID, RegionProtection and MiscSR features. Implement instructions for Exception, Interrupt and RegionProtection features with tests.
- [ ] Implement Atomic operations support.

Clang:
- [ ] [Add the Xtensa target.](https://github.com/llvm/llvm-project/pull/118008)
- [ ] Implement Xtensa ABI lowering.
- [ ] Add subtargets ESP32. ESP8266 and ESP32-S2.
- [ ] Implement ESP Baremetal toolchain

**Comment by andreisfr** [comment]
The Xtensa backend upstreaming status update:
Xtensa features and toolchain implementation
LLVM:


- [ ]  https://github.com/llvm/llvm-project/pull/119639
- [ ]  https://github.com/llvm/llvm-project/pull/121118)
- [ ]  Implement Xtensa FP option and operations
- [ ]  Implement Loop, SEXT and NSA features.
- [ ]  Implement Mul32, Mul32High and Div32 features.
- [ ]  Implement Mac16 feature and operations.
- [ ]  Implement Debug, DFPAccel, S32C1I, THREADPTR, Extended L32R, ATOMCTL, MEMCTL features.
- [ ]  Implement Exception, HighPriInterrupts, Coprocessor, Interrupt, RelocatableVector, TimerInt, PRID, RegionProtection and MiscSR features. Implement instructions for Exception, Interrupt and RegionProtection features with tests.
- [ ]  Implement Atomic operations support.


Clang:


- [ ]  https://github.com/llvm/llvm-project/pull/118008
- [ ]  Implement Xtensa ABI lowering.
- [ ]  Add subtargets ESP32. ESP8266 and ESP32-S2.
- [ ]  Implement ESP Baremetal toolchain

**Comment by andreisfr** [comment]
We finished first step with adding basic Xtensa backend functionality to upstream. And our plans is to publish next PRs mostly in parallel (2-3 PR simultaneously if possible), so it is expected  speedup of the upstreaming process.

**Comment by andreisfr** [comment]
Hi guys! Update of the Xtensa backend upstreaming status:
Xtensa features and toolchain implementation
LLVM:

- [ ] https://github.com/llvm/llvm-project/pull/119639
- [ ] https://github.com/llvm/llvm-project/pull/121118)
- [ ] Implement Loop, SEXT and NSA features.
- [ ] Implement Mul32, Mul32High and Div32 features.
- [ ] Implement Mac16 feature and operations.
- [ ] Implement Xtensa FP option and operations. (PR in review) https://github.com/llvm/llvm-project/pull/136086
- [ ] Implement S32C1I Option,  and Atomic operations support. (PR in review) https://github.com/llvm/llvm-project/pull/137134
- [ ] Implement RelocatableVector, RegionProtection , MiscSR , MEMCTL , Extended L32R. (PR in review) https://github.com/llvm/llvm-project/pull/137135
- [ ] Implement Exception, HighPriInterrupts, Coprocessor, Interrupt, TimerInt, PRID, Implement instructions for Exception, Interrupt  features with tests.
- [ ] Implement Debug, DFPAccel, THREADPTR features.

Clang:

- [ ] https://github.com/llvm/llvm-project/pull/118008
- [ ] Implement Xtensa ABI lowering.
- [ ] Add subtargets ESP32. ESP8266 and ESP32-S2.
- [ ] Implement ESP Baremetal toolchain

**Comment by Be-ing** [comment]
> Also we work on support of the Xtensa architecture in other LLVM projects (lld, compiler-rt, etc)

I would appreciate Xtensa support in lld. The GNU linker is so slow.

**Comment by JetForMe** [comment]
> everybody who is blocked due to lack of Xtensa (Espressif chips) support in upstream

I'm looking to use Swift for development. I can do that on C6, which is great for my Matter devices, but I use S3 for a lot of other things.

**Comment by TurkeyMan** [comment]
I've been trying out what's already been upstream-ed; I can build my project, but if I add -O1 (any optimisation level), I suddenly see ICE surrounding exception handling.
I asked Claude to try and diagnose the ICE, and he determined that exception handling is basically just not implemented... is this the case?

> Any function with an invoke + landingpad will trigger this. The key pattern is: invoke that can throw → landing pad with catch clause → resume path.
> 
> MIR Evidence
> The landing pad block in the crashing function shows undefined registers:
> 
> 
> bb.11 (%ir-block.53, landing-pad):
>   EH_LABEL <mcsymbol >
>   %17:ar = COPY %33:ar     ← %33 never defined anywhere
>   BEQI %34:ar, 1, %bb.9    ← %34 never defined anywhere
> Compare to working backends (x86, ARM) where %33/%34 would be defined by the ISel lowering of the landingpad instruction's {ptr, i32} result.
> 
> Likely Fix Location
> The Xtensa ISel lowering likely needs to handle ISD::EH_RETURN / exception handling register setup in XtensaTargetLowering::LowerOperation() or similar, defining the exception pointer and selector registers in the landing pad entry. Most backends do this via getExceptionPointerRegister() / getExceptionSelectorRegister() in their TargetLowering class.

**Comment by aykevl** [comment]
For TinyGo we're currently using the Espressif fork of LLVM, but this makes LLVM updates harder.
What we would need to switch to upstream LLVM is support for a few common chips (like esp8266 and esp32), support for Clang (for building C libraries like compiler-rt and picolibc) and support for LLD (for linking the resulting binary). Unfortunately, that means we need almost the entirety of Xtensa support upstreamed for it to be usable :/
